### ISSUE TITLE
pcolormesh and broadcasting

### ISSUE THREAD

**Issue by henryiii** [matplotlib/matplotlib#15840]
Currently, pcolormesh does not accept "sparse" arrays of the form (N,1) x (1,M), but only fully broadcast ones, for X and Y. So the following code works:

```python
X, Y = np.mgrid[-10:11, -10:11]
C = X**2 + Y**2
plt.pcolormesh(X, Y, C)
```

But the following code throws an error:

```python
X, Y = np.ogrid[-10:11, -10:11]
C = X**2 + Y**2
plt.pcolormesh(X, Y, C)
```

<details><summary>Error (click to expand)</summary>

```text
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-160-e4c56d938633> in <module>
      1 X, Y = np.ogrid[-10:11, -10:11]
      2 C = X**2 + Y**2
----> 3 plt.pcolormesh(X, Y, C)

/usr/local/miniconda3/envs/myhist/lib/python3.7/site-packages/matplotlib/pyplot.py in pcolormesh(alpha, norm, cmap, vmin, vmax, shading, antialiased, data, *args, **kwargs)
   2749         *args, alpha=alpha, norm=norm, cmap=cmap, vmin=vmin,
   2750         vmax=vmax, shading=shading, antialiased=antialiased,
-> 2751         **({"data": data} if data is not None else {}), **kwargs)
   2752     sci(__ret)
   2753     return __ret

/usr/local/miniconda3/envs/myhist/lib/python3.7/site-packages/matplotlib/__init__.py in inner(ax, data, *args, **kwargs)
   1597     def inner(ax, *args, data=None, **kwargs):
   1598         if data is None:
-> 1599             return func(ax, *map(sanitize_sequence, args), **kwargs)
   1600 
   1601         bound = new_sig.bind(ax, *args, **kwargs)

/usr/local/miniconda3/envs/myhist/lib/python3.7/site-packages/matplotlib/axes/_axes.py in pcolormesh(self, alpha, norm, cmap, vmin, vmax, shading, antialiased, *args, **kwargs)
   6163         allmatch = (shading == 'gouraud')
   6164 
-> 6165         X, Y, C = self._pcolorargs('pcolormesh', *args, allmatch=allmatch)
   6166         Ny, Nx = X.shape
   6167         X = X.ravel()

/usr/local/miniconda3/envs/myhist/lib/python3.7/site-packages/matplotlib/axes/_axes.py in _pcolorargs(funcname, allmatch, *args)
   5748             raise TypeError(
   5749                 'Incompatible X, Y inputs to %s; see help(%s)' % (
-> 5750                 funcname, funcname))
   5751         if allmatch:
   5752             if (Nx, Ny) != (numCols, numRows):

TypeError: Incompatible X, Y inputs to pcolormesh; see help(pcolormesh)
```
</details>

Requiring the following fix:

```python
X, Y = np.ogrid[-10:11, -10:11]
C = X**2 + Y**2
plt.pcolormesh(*np.broadcast_arrays(X, Y), C)
# or
plt.pcolormesh(X.flatten(),  Y.flatten(), C)
```

(In my actual code, I have an API that produces the sparse arrays and it makes `pcolormesh` harder to use)

Would you be interested in a PR that allows correctly formed 2D input instead of failing? I could probably do it fairly easily if it is of interest.

(As a broadcasting reminder, the this is for X = (N,1), Y=(1,M), C=(N,M), it would not affect 1D arrays/lists, and 2D or X = (1, N), Y=(M,1), C=(N,M) already works )

**Comment by anntzer** [comment]
Seems fine to me, but I think the correct shapes are (1, N) and (M, 1)?

**Comment by jklymak** [comment]
You’d presumably have to do this wevrywhere X and Y are allowed to be matrices.  If you are able to detect sparse arrays like this I would also not broadcast them but rather turn them into two vectors.

**Comment by henryiii** [comment]
Thanks @jklymak, I've updated the description above. @anntzer, I think I got the shapes right:

```python
>>> arrs = np.ogrid[:2, :3, :4]
>>> print(*(a.shape for a in arrs))
(2, 1, 1) (1, 3, 1) (1, 1, 4)
```

**Comment by henryiii** [comment]
@jklymak, what do you mean by everywhere? In other functions, in this function, etc?

**Comment by jklymak** [comment]
There are lots of other methods where X, Y are accepted as matrices.  i.e. `quiver`, the other pcolormesh variations, etc.

**Comment by anntzer** [comment]
Your original example doesn't work if X and Y have different lengths, you need to do `pcolormesh(Y, X, ...)` (or really rather `pcolormesh(X, Y, C.T)`).

**Comment by henryiii** [comment]
`quiver` seems to do the wrong thing:

```python
import matplotlib.pyplot as plt
import numpy as np

X = np.arange(-10, 10, 1)
Y = np.arange(-10, 10, 1)
U, V = np.meshgrid(X, Y, sparse=True)

fig, ax = plt.subplots()
q = ax.quiver(X, Y, U, V)
ax.quiverkey(q, X=0.3, Y=1.1, U=10,
              label='Quiver key, length = 10', labelpos='E')

plt.show()
```

Does not throw an error, but produces the wrong plot:

![quiver_sparse](https://user-images.githubusercontent.com/4616906/70334636-c95d4680-1813-11ea-821d-37dcad9660f5.png)


Instead of:

![quiver](https://user-images.githubusercontent.com/4616906/70334646-cd896400-1813-11ea-90bf-db0e8e317afe.png)


(if `sparse=True` is removed)

**Comment by henryiii** [comment]
@anntzer, you are correct, ij vs. xy indexing got me. My original code did have a `.T` in it, but that was lost when I made the example.

**Comment by anntzer** [comment]
That's because X and Y have the right (non-broadcasted) shapes in that case?

**Comment by henryiii** [comment]
Yes, `(10, 21)`.

![image](https://user-images.githubusercontent.com/4616906/70355248-a8aae600-183f-11ea-834e-c9f455cf19c3.png)

So

```python
X, Y = np.ogrid[-5:5, -10:11]
C = X**2 + Y**2
plt.pcolormesh(X, Y, C)
```

Should work identically, right?

**Comment by henryiii** [comment]
Ah, I see what you mean. Let me play with it for a bit.

**Comment by henryiii** [comment]
So I think the problem is that only one direction works, that is:

```python
plt.pcolormesh(X, Y, Z)
```

Then

```python
plt.pcolormesh(X.T, Y.T, Z.T)
```

also works, but *only* if the arrays already have been broadcast.

**Comment by henryiii** [comment]
(Fixing `quiver` might also be a good idea, since it currently seems to flatten the input, but I could be wrong there)

**Comment by jklymak** [comment]
Is there a reason to reopen this?  I am pretty sure our broadcasting rules are ok. But maybe I am misunderstanding the issue?

**Comment by tacaswell** [comment]
Re-reading it was not clear to me we had actually come to that conclusion, just that we had stopped talking about it.

**Comment by jklymak** [comment]
I'd close and let the OP open a new issue if they are still having issues.  The problem changed a couple of times during discussion.

**Comment by henryiii** [comment]
I don’t think it changed. The key issue is broadcastable matrices should behave like fully broadcast matrices.

**Comment by jklymak** [comment]
> I don’t think it changed. The key issue is broadcastable matrices should behave like fully broadcast matrices.

You mean something specific by that, but it's not clear from this issue what that is.  

As noted above: 

> Your original example doesn't work if X and Y have different lengths, you need to do `pcolormesh(Y, X, ...)` (or really rather `pcolormesh(X, Y, C.T)`).

you just have the dimensions backwards.  If your contention is that we should transpose C for the user in that case, I'd be pretty strongly against that proposal.  Data shape is a pretty strong check that you understand your data structure, and just swapping dimensions for the user is a bad idea.
